src/pygeoprocessing/multiprocessing/raster_calculator.py

```diff
@@ -0,0 +1,589 @@
+            defaults to ``multiprocessing.cpu_count()``.
+        calc_raster_stats (boolean): If True, calculates and sets raster
+            statistics (min, max, mean, and stdev) for target raster.
+        largest_block (int): Attempts to internally iterate over raster blocks
+            with this many elements.  Useful in cases where the blocksize is
+            relatively small, memory is available, and the function call
+            overhead dominates the iteration.  Defaults to 2**20.  A value of
+            anything less than the original blocksize of the raster will
+            result in blocksizes equal to the original size.
+        raster_driver_creation_tuple (tuple): a tuple containing a GDAL driver
+            name string as the first element and a GDAL creation options
+            tuple/list as the second. Defaults to
+            geoprocessing.DEFAULT_GTIFF_CREATION_TUPLE_OPTIONS.
+
+    Returns:
+        None
+
+    Raises:
+        ValueError: invalid input provided
+
+    """
+    raster_info_list = [
+        get_raster_info(path_band[0])
+        for path_band in base_raster_path_band_const_list
+        if _is_raster_path_band_formatted(path_band)]
+
+    _validate_raster_input(
+        base_raster_path_band_const_list, raster_info_list, target_raster_path)
+
+    n_cols, n_rows = _calculate_target_raster_size(
+        raster_info_list, base_raster_path_band_const_list)
+
+    # create a "canonical" argument list that contains only
+    # (file paths, band id) tuples, 2d numpy arrays, or raw values
+    base_canonical_arg_list = []
+    for value in base_raster_path_band_const_list:
+        # the input has been tested and value is either a raster/path band
+        # tuple, 1d ndarray, 2d ndarray, or (value, 'raw') tuple.
+        if _is_raster_path_band_formatted(value):
+            # it's a raster/path band, keep track of open raster and band
+            # for later so we can __swig_destroy__ them.
+            base_canonical_arg_list.append(
+                RasterPathBand(value[0], value[1]))
+        elif isinstance(value, numpy.ndarray):
+            if value.ndim == 1:
+                # easier to process as a 2d array for writing to band
+                base_canonical_arg_list.append(
+                    value.reshape((1, value.shape[0])))
+            else:  # dimensions are two because we checked earlier.
+                base_canonical_arg_list.append(value)
+        else:
+            # it's a regular tuple
+            base_canonical_arg_list.append(value)
+
+    if datatype_target not in _VALID_GDAL_TYPES:
+        raise ValueError(
+            'Invalid target type, should be a gdal.GDT_* type, received '
+            '"%s"' % datatype_target)
+
+    # create target raster
+    raster_driver = gdal.GetDriverByName(raster_driver_creation_tuple[0])
+    try:
+        os.makedirs(os.path.dirname(target_raster_path))
+    except OSError:
```

Comment:
Could you describe a circumstancs where we'd have an `OSError` that won't cause the rest of this `raster_calculator` call to crash?

Reply:
It's such a habit to write that pair I don't think about it anymore! I added an explicit check for the kind of exception in 65abdd59dd126c5273f321898de08c423bc6fe28